### PR TITLE
8316814: NMT: A separate script or Java program is needed to analyze and make useful reports ouf of JMH benchmarks outputs.

### PR DESCRIPTION
**Pull request by afshin-zafari**
It reads the benchmark report file whose path is given as command line. Then extract percentage of overhead of NMT modes for different methods used in benchmarks.
Locally tested with some sample report files.
From an input benchmark report like this: 
```
# JMH version: 1.37
# VM version: JDK 22-internal, OpenJDK 64-Bit Server VM,......in...
# VM invoker:.....
# VM options: -XX:NativeMemoryTracking=detail --add-exports java.base/jdk.internal.misc=ALL-UNNAMED -Djava.library.path=...
# Blackhole mode: compiler (auto-detected, use -Djmh.blackhole.autoDetect=false to disable)
# Warmup: 2 iterations, 1 s each
# Measurement: 5 iterations, 1 s each
# Timeout: 10 min per iteration
# Threads: 1 thread, will synchronize iterations
# Benchmark mode: Average time, time/op
# Benchmark: org.openjdk.bench.vm.runtime.NMTBenchmark.NMTDetail.mixAallocateFreeMemory
# Parameters: (N = 100000, THREADS = 4)

.
.
.


Benchmark                                               (N)  (THREADS)  Mode  Cnt    Score    Error  Units
NMTBenchmark.NMTDetail.mixAallocateFreeMemory        100000          0  avgt   10   87.010 ? 10.630  ms/op
NMTBenchmark.NMTDetail.mixAallocateFreeMemory        100000          4  avgt   10    2.473 ?  0.728  ms/op
NMTBenchmark.NMTDetail.mixAllocateReallocateMemory   100000          0  avgt   10  225.054 ?  4.381  ms/op
NMTBenchmark.NMTDetail.mixAllocateReallocateMemory   100000          4  avgt   10    2.247 ?  0.599  ms/op
NMTBenchmark.NMTDetail.onlyAllocateMemory            100000          0  avgt   10  147.346 ?  8.226  ms/op
NMTBenchmark.NMTDetail.onlyAllocateMemory            100000          4  avgt   10    2.285 ?  0.611  ms/op
NMTBenchmark.NMTOff.mixAallocateFreeMemory           100000          0  avgt   10   40.405 ?  8.115  ms/op
NMTBenchmark.NMTOff.mixAallocateFreeMemory           100000          4  avgt   10    2.299 ?  0.658  ms/op
NMTBenchmark.NMTOff.mixAllocateReallocateMemory      100000          0  avgt   10  158.700 ? 10.750  ms/op
NMTBenchmark.NMTOff.mixAllocateReallocateMemory      100000          4  avgt   10    2.282 ?  0.723  ms/op
NMTBenchmark.NMTOff.onlyAllocateMemory               100000          0  avgt   10  113.949 ? 21.257  ms/op
NMTBenchmark.NMTOff.onlyAllocateMemory               100000          4  avgt   10    2.155 ?  0.682  ms/op
NMTBenchmark.NMTSummary.mixAallocateFreeMemory       100000          0  avgt   10   50.436 ?  4.576  ms/op
NMTBenchmark.NMTSummary.mixAallocateFreeMemory       100000          4  avgt   10    2.192 ?  0.737  ms/op
NMTBenchmark.NMTSummary.mixAllocateReallocateMemory  100000          0  avgt   10  170.408 ?  4.556  ms/op
NMTBenchmark.NMTSummary.mixAllocateReallocateMemory  100000          4  avgt   10    2.344 ?  0.645  ms/op
NMTBenchmark.NMTSummary.onlyAllocateMemory           100000          0  avgt   10  117.480 ?  8.846  ms/op
NMTBenchmark.NMTSummary.onlyAllocateMemory           100000          4  avgt   10    2.367 ?  0.694  ms/op
```

it creates summary like this:
```

                                  Method Threads   Off  Summary Detail
                  mixAallocateFreeMemory     0     40.41  50.44   87.01
                                                          24.83% 115.34%
                  mixAallocateFreeMemory     4      2.30   2.19    2.47
                                                          -4.65%   7.57%
             mixAllocateReallocateMemory     0    158.70 170.41  225.05
                                                           7.38%  41.81%
             mixAllocateReallocateMemory     4      2.28   2.34    2.25
                                                           2.72%  -1.53%
                      onlyAllocateMemory     0    113.95 117.48  147.35
                                                           3.10%  29.31%
                      onlyAllocateMemory     4      2.16   2.37    2.29
                                                           9.84%   6.03%
```

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8316814](https://bugs.openjdk.org/browse/JDK-8316814): NMT: A separate script or Java program is needed to analyze and make useful reports ouf of JMH benchmarks outputs. (**Enhancement** - P4)

### Reviewers
 * [Johan Sjölen](https://openjdk.org/census#jsjolen) (@jdksjolen - **Reviewer**) ⚠️ Review applies to [bbe147ab](https://git.openjdk.org/jdk/pull/16070/files/bbe147ab6cfa8bdc7172ea4e5dfc8fcaa135c8a3)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/16070/head:pull/16070` \
`$ git checkout pull/16070`

Update a local copy of the PR: \
`$ git checkout pull/16070` \
`$ git pull https://git.openjdk.org/jdk.git pull/16070/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 16070`

View PR using the GUI difftool: \
`$ git pr show -t 16070`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/16070.diff">https://git.openjdk.org/jdk/pull/16070.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/16070#issuecomment-1772395666)

### CODE REVIEW

**Comment by gerard-ziemski** [comment]
Can we add error report please?

**Review by gerard-ziemski — CHANGES_REQUESTED** [review]
Can we add error values to the report please?

**Comment by gerard-ziemski** [comment]
Would a layout like this be easier to parse (everything on one line)

                                  Method Threads   Off    Summary Diff     Detail Diff
                  mixAallocateFreeMemory     0     40.41    50.44 24.83%   87.01  115.34%
                  mixAallocateFreeMemory     4      2.30     2.19 -4.65%   2.47   7.57%
             mixAllocateReallocateMemory     0    158.70   170.41 7.38%    225.05 41.81%
             mixAllocateReallocateMemory     4      2.28     2.34 2.72%    2.25   -1.53%
                      onlyAllocateMemory     0    113.95   117.48 3.10%    147.35 29.31%
                      onlyAllocateMemory     4      2.16     2.37 9.84%    2.29   6.03%

I'm not sure myself yet which one I prefer, they both look a bit busy, but this one require less vertical scanning, but more horizontal scanning...

**Comment by gerard-ziemski** [comment]
I think we might need 3 decimal places to show to the user (like the original output).

When I try the numbers as shown say here:

                  mixAallocateFreeMemory     4      2.30   2.19    2.47
                                                          -4.65%   7.57%
  
20.19 - 2.30 == -0.11 --> -0.11 / 2.30 --> -0.047826086956522 * 100 --> -4.78%

I get quite a different number --> -4.78% vs -4.65%

**Comment by afshin-zafari** [comment]
> Would a layout like this be easier to parse (everything on one line)
> 
> ```
>                               Method Threads   Off    Summary Diff     Detail Diff
>               mixAallocateFreeMemory     0     40.41    50.44 24.83%   87.01  115.34%
>               mixAallocateFreeMemory     4      2.30     2.19 -4.65%   2.47   7.57%
>          mixAllocateReallocateMemory     0    158.70   170.41 7.38%    225.05 41.81%
>          mixAllocateReallocateMemory     4      2.28     2.34 2.72%    2.25   -1.53%
>                   onlyAllocateMemory     0    113.95   117.48 3.10%    147.35 29.31%
>                   onlyAllocateMemory     4      2.16     2.37 9.84%    2.29   6.03%
> ```
> 
> I'm not sure myself yet which one I prefer, they both look a bit busy, but this one requires less vertical scanning, but more horizontal scanning...

Fixed.

**Comment by afshin-zafari** [comment]
> I think we might need 3 decimal places to show to the user (like the original output).
> 
> When I try the numbers as shown say here:
> 
> ```
>               mixAallocateFreeMemory     4      2.30   2.19    2.47
>                                                       -4.65%   7.57%
> ```
> 
> 20.19 - 2.30 == -0.11 --> -0.11 / 2.30 --> -0.047826086956522 * 100 --> -4.78%
> 
> I get quite a different number --> -4.78% vs -4.65%

Excellent catch! Numbers shown with 3 decimal digits.

File: test/micro/org/openjdk/bench/vm/runtime/NMTAnalyze.java [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+  }
+  public static BufferedReader reader;
+  public static String fileName;
+  public static List<BenchmarkRecord> records = new ArrayList<>();
+  private static void skipUntilReport() {
+    try {
+      String line = reader.readLine();
+      while (line != null) {
+        String ReportStartKeyword = "Benchmark";
+        if (!line.startsWith(ReportStartKeyword) || line.contains("#")) {
```
Comment (jdksjolen) [inline test/micro/org/openjdk/bench/vm/runtime/NMTAnalyze.java:61]: Are there lines which starts with `Benchmark` but contains a `#` that we need to skip?

File: test/micro/org/openjdk/bench/vm/runtime/NMTAnalyze.java [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+        } else {
+          break;
+        }
+      }
+    } catch (IOException ioe) {
+      System.out.println(ioe.getMessage());
+    }
+  }
+  private static void readRecords() {
+    try {
```
Comment (jdksjolen) [inline test/micro/org/openjdk/bench/vm/runtime/NMTAnalyze.java:73]: Same thing here regarding `IOException` should be thrown

File: test/micro/org/openjdk/bench/vm/runtime/NMTAnalyze.java [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+      do {
```
Comment (jdksjolen) [inline test/micro/org/openjdk/bench/vm/runtime/NMTAnalyze.java:74]: Why not the same loop pattern as in `skipUntilReport`?

File: test/micro/org/openjdk/bench/vm/runtime/NMTAnalyze.java [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+        String line = reader.readLine();
+        if (line == null) break;
+        if (!line.startsWith("NMTBenchmark"))
+          continue;
+        String[] _fields= line.split(" ");
+        BenchmarkRecord rec = new BenchmarkRecord();
+        List<String> __fields = Arrays.asList(_fields);
+        String[] fields = __fields.stream().filter(f -> f.length() > 0).collect(Collectors.toList()).toArray(new String[0]);
```
Comment (jdksjolen) [inline test/micro/org/openjdk/bench/vm/runtime/NMTAnalyze.java:82]: Could fold the fields handling into one line without loss of readability: `Arrays.asList(line.split(" ")).stream()`

**Review by jdksjolen — COMMENTED** [review]
Hi Afshin!

PR output looks good, but I've got some comments on the code.

File: test/micro/org/openjdk/bench/vm/runtime/NMTAnalyze.java (hunk 1 above)
Comment (afshin-zafari) [reply test/micro/org/openjdk/bench/vm/runtime/NMTAnalyze.java:61]: Yes, there are. for each run there is a part like this. with `# Benchmark`:
```
# JMH version: 1.37
# VM version: JDK 22-internal, OpenJDK 64-Bit Server VM,......in...
# VM invoker:.....
# VM options: -XX:NativeMemoryTracking=detail --add-exports java.base/jdk.internal.misc=ALL-UNNAMED -Djava.library.path=...
# Blackhole mode: compiler (auto-detected, use -Djmh.blackhole.autoDetect=false to disable)
# Warmup: 2 iterations, 1 s each
# Measurement: 5 iterations, 1 s each
# Timeout: 10 min per iteration
# Threads: 1 thread, will synchronize iterations
# Benchmark mode: Average time, time/op
# Benchmark: org.openjdk.bench.vm.runtime.NMTBenchmark.NMTDetail.mixAallocateFreeMemory
# Parameters: (N = 100000, THREADS = 4)
```

File: test/micro/org/openjdk/bench/vm/runtime/NMTAnalyze.java [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+
+class NMTAnalyze {
+  public static class BenchmarkRecord {
+    public String jmh_method;
+    public int N, threads, cnt;
+    public double score, error;
+    public String mode;
+    public String units;
+  }
+  public static class BenchmarkFields {
```
[the remark this answers is not in the document]
Comment (afshin-zafari) [reply test/micro/org/openjdk/bench/vm/runtime/NMTAnalyze.java:42]: Done.

File: test/micro/org/openjdk/bench/vm/runtime/NMTAnalyze.java [hunk 6]
```diff
@@ -0,0 +1,153 @@
...
+    static int CNT           = 4;
+    static int SCORE         = 5;
+    static int QUESTION_MARK = 6;
+    static int ERROR         = 7;
+    static int UNITS         = 8;
+  }
+  public static BufferedReader reader;
+  public static String fileName;
+  public static List<BenchmarkRecord> records = new ArrayList<>();
+  private static void skipUntilReport() {
```
[the remark this answers is not in the document]
Comment (afshin-zafari) [reply test/micro/org/openjdk/bench/vm/runtime/NMTAnalyze.java:56]: Fixed.

File: test/micro/org/openjdk/bench/vm/runtime/NMTAnalyze.java [hunk 7]
```diff
@@ -0,0 +1,153 @@
...
+        if (!line.startsWith(ReportStartKeyword) || line.contains("#")) {
+          line = reader.readLine();
+          continue;
+        } else {
+          break;
+        }
+      }
+    } catch (IOException ioe) {
+      System.out.println(ioe.getMessage());
+    }
```
[the remark this answers is not in the document]
Comment (afshin-zafari) [reply test/micro/org/openjdk/bench/vm/runtime/NMTAnalyze.java:70]: Fixed.

File: test/micro/org/openjdk/bench/vm/runtime/NMTAnalyze.java (hunk 2 above)
Comment (afshin-zafari) [reply test/micro/org/openjdk/bench/vm/runtime/NMTAnalyze.java:73]: Fixed.

File: test/micro/org/openjdk/bench/vm/runtime/NMTAnalyze.java (hunk 3 above)
Comment (afshin-zafari) [reply test/micro/org/openjdk/bench/vm/runtime/NMTAnalyze.java:74]: Done.

File: test/micro/org/openjdk/bench/vm/runtime/NMTAnalyze.java (hunk 4 above)
Comment (afshin-zafari) [reply test/micro/org/openjdk/bench/vm/runtime/NMTAnalyze.java:82]: Done.

File: test/micro/org/openjdk/bench/vm/runtime/NMTAnalyze.java [hunk 8]
```diff
@@ -0,0 +1,153 @@
...
+      readRecords();
+      analyzeRecords();
+			reader.close();
+		} catch (IOException ioe) {
+      ioe.printStackTrace();
+      System.out.println(ioe.getMessage());
+		} finally {
+    }
+
+  }
```
[the remark this answers is not in the document]
Comment (afshin-zafari) [reply test/micro/org/openjdk/bench/vm/runtime/NMTAnalyze.java:152]: Done.

**Review by afshin-zafari — COMMENTED** [review]
Thanks for the comments.

**Comment by gerard-ziemski** [comment]
Just noticed a small typo in mixAallocateFreeMemory, it should be mixAllocateFreeMemory I think.

**Comment by afshin-zafari** [comment]
> Just noticed a small typo in mixAallocateFreeMemory, it should be mixAllocateFreeMemory I think.

Yes, unfortunately the typo is in the JMH tests method and cannot be corrected here (because the report are generated with this typo) until it is changed in the original tests.

**Comment by afshin-zafari** [comment]
> Just noticed a small typo in mixAallocateFreeMemory, it should be mixAllocateFreeMemory I think.

Unfortunately the typo is in the original JMH test. Since that change is not yet merged, the typo cannot be fixed in the test now.

**Comment by afshin-zafari** [comment]
@jdksjolen and @gerard-ziemski ,
Is there any other comments on this PR?

**Review by jdksjolen — APPROVED** [review]
Hi Afshin, thanks for the ping. Changes looks good! Could you fix the whitespace issues that the bot has commented on?

File: test/micro/org/openjdk/bench/vm/runtime/NMTAnalyze.java [hunk 9]
```diff
@@ -0,0 +1,152 @@
...
+      skipUntilReport(reader);
+      readRecords(reader);
+      analyzeRecords(reader);
+
+      reader.close();
+    } catch (IOException ioe) {
+      ioe.printStackTrace();
+      System.out.println(ioe.getMessage());
+    } finally {
+    }
```
Comment (dholmes-ora) [inline test/micro/org/openjdk/bench/vm/runtime/NMTAnalyze.java:149]: This serves no purpose.

**Review by dholmes-ora — COMMENTED** [review]
Not a review but as per my comment in JBS I'm unclear where this tool really belongs. Perhaps @shipilev  has an opinion?

File: test/micro/org/openjdk/bench/vm/runtime/NMTAnalyze.java (hunk 9 above)
Comment (afshin-zafari) [reply test/micro/org/openjdk/bench/vm/runtime/NMTAnalyze.java]: Fixed.

**Comment by afshin-zafari** [comment]
> Not a review but as per my comment in JBS I'm unclear where this tool really belongs. Perhaps @shipilev has an opinion?

As a reply to your comment, I moved the test to test/micro/bench/org/openjdk/bench/vm/runtime. It was already in ../bench/java/util. Maybe I misunderstood your comment.

**Comment by afshin-zafari** [comment]
> b) Parsing the human-readable result is a proper way to do this, as it is not guaranteed to be stable; JMH has machine-readable JSON output for external tools;

How to process JSON files in JDK?

**Comment by afshin-zafari** [comment]
@shipilev  and @dholmes-ora 
I think since this tool (or even any other tool) is completely related/dependent to a specific benchmark, it should be put next to the benchmark. If we put it somewhere else, we should mention somehow which benchmark it is related to. There is a high dependency between the benchmark and the tool. Whenever the benchmark changes, the tool should be changed accordingly, since it relies on the names and formats used in the benchmark.

**Comment by dholmes-ora** [comment]
Sorry @afshin-zafari  I'm really not sure what to do about this one, but it seems a JMH issue.

**Comment by afshin-zafari** [comment]
Since this PR works only for one JMH benchmark and is not general, it cannot be introduced in JVM.
So, I close it with no merge.

**Pull request closed without merging**